# Data cleaning notebook

This notebook contains the examples from chapter 2 of the book. Let us start by reading in the data into a data frame named `bike`.

In [19]:
#install.packages("tidyverse")
#install.packages("DataCombine")

package 'DataCombine' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\MaZeeT\AppData\Local\Temp\RtmpWoYar5\downloaded_packages


In [2]:
bike <- read.csv("https://raw.githubusercontent.com/jgendron/com.packtpub.intro.r.bi/master/Chapter2-DataCleaning/data/Ch2_raw_bikeshare_data.csv", stringsAsFactors = FALSE)

## 1 Summarizing your data for inspection

We will start by looking a bit at the data

In [3]:
str(bike)

'data.frame':	17379 obs. of  13 variables:
 $ datetime  : chr  "1/1/2011 0:00" "1/1/2011 1:00" "1/1/2011 2:00" "1/1/2011 3:00" ...
 $ season    : int  1 1 1 1 1 1 1 1 1 1 ...
 $ holiday   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ workingday: int  0 0 0 0 0 0 0 0 0 0 ...
 $ weather   : int  1 1 1 1 1 2 1 1 1 1 ...
 $ temp      : num  9.84 9.02 9.02 9.84 9.84 ...
 $ atemp     : num  14.4 13.6 13.6 14.4 14.4 ...
 $ humidity  : chr  "81" "80" "80" "75" ...
 $ windspeed : num  0 0 0 0 0 ...
 $ casual    : int  3 8 5 3 0 0 2 1 1 8 ...
 $ registered: int  13 32 27 10 1 1 0 2 7 6 ...
 $ count     : int  16 40 32 13 1 1 2 3 8 14 ...
 $ sources   : chr  "ad campaign" "www.yahoo.com" "www.google.fi" "AD campaign" ...


Clear problems that we can note already are that the `datetime` column is in characters and not a proper date-time format in R. Moreover, the `humidity` variable is also character, while it looks like it should be numeric or integer.

In [4]:
dim(bike)
head(bike)
tail(bike)

[1] 17379    13

datetime      season holiday workingday weather temp atemp  humidity
1 1/1/2011 0:00 1      0       0          1       9.84 14.395 81      
2 1/1/2011 1:00 1      0       0          1       9.02 13.635 80      
3 1/1/2011 2:00 1      0       0          1       9.02 13.635 80      
4 1/1/2011 3:00 1      0       0          1       9.84 14.395 75      
5 1/1/2011 4:00 1      0       0          1       9.84 14.395 75      
6 1/1/2011 5:00 1      0       0          2       9.84 12.880 75      
  windspeed casual registered count sources      
1 0.0000    3      13         16    ad campaign  
2 0.0000    8      32         40    www.yahoo.com
3 0.0000    5      27         32    www.google.fi
4 0.0000    3      10         13    AD campaign  
5 0.0000    0       1          1    Twitter      
6 6.0032    0       1          1    www.bing.com

datetime         season holiday workingday weather temp  atemp  humidity
17374 12/31/2012 18:00 1      0       1          2       10.66 13.635 48      
17375 12/31/2012 19:00 1      0       1          2       10.66 12.880 60      
17376 12/31/2012 20:00 1      0       1          2       10.66 12.880 60      
17377 12/31/2012 21:00 1      0       1          1       10.66 12.880 60      
17378 12/31/2012 22:00 1      0       1          1       10.66 13.635 56      
17379 12/31/2012 23:00 1      0       1          1       10.66 13.635 65      
      windspeed casual registered count sources      
17374  8.9981   10     326        336   facebook page
17375 11.0014    5     206        211   NA           
17376 11.0014    4     140        144   AD campaign  
17377 11.0014    3      96         99   AD campaign  
17378  8.9981    4      90         94   ad campaign  
17379  8.9981    3      50         53   direct

## 2 Finding and fixing flawed data

We will now try to find and fix errors in the data. Note that, this is detective work and there is not always one answer and one method that find and solves all errors.

### Missing values

Missing values is a common problem occuring with most data sets. Sometimes, a specific values is used to indicate that the value is missing. In R this is represented by the value `NA`. In the above `tail` call you can see such a case in the last column. First, let us see how to count the number of `NA` in a data frame:

In [5]:
table(is.na(bike))


 FALSE   TRUE 
225373    554 

This show us that our data frame `bike` contains 554 `NA` values, while 225373 values are not missing. The following code can help us see if it is all variables that contain `NA`:

In [6]:
library(stringr)
str_detect(bike, "NA")

Warning message in stri_detect_regex(string, pattern, negate = negate, opts_regex = opts(pattern)):
"argument is not an atomic vector; coercing"

[1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[13]  TRUE

This function returns `TRUE` or `FALSE` for each column in our data frame. The result shows that it is only the last column that contains `NA` values. Another way to confirm this is by:

In [7]:
table(is.na(bike$sources))


FALSE  TRUE 
16825   554 

Here we can see that there as many `NA` values in the `sources` column as there is in the entire data frame. We will fix these missing values later.

Dealing with missing values is a huge topic on its own! We will return to this next time also.

### Erroneous values

We now return to the issue with the `humidity` attribute. We will start by searching for characters in the column:

In [8]:
bad_data <- str_subset(bike$humidity, "[a-z A-Z]")
bad_data

[1] "x61"

From this we can see that the value `x61` appears somewhere in the column `humidity`. This is clearly an error. There are not always a clear answer or solution, but in this case, it seems like someone just miss-typed `61` as `x61`. Let us find the location of this error:

In [9]:
location <- str_detect(bike$humidity, bad_data)
bike[location, ]

datetime        season holiday workingday weather temp  atemp humidity
14177 8/18/2012 21:00 3      0       0          1       27.06 31.06 x61     
      windspeed casual registered count sources     
14177 0         90     248        338   www.bing.com

Note, that the `str_detect` function give us a vector of `TRUE` and `FAlSE` which is only true for the row that has the error. This we can use to subset the `bike` data frame to see the row with the error.

We can now replace this error in the following way and inspect that we fixed the error:

In [10]:
bike$humidity <- str_replace_all(bike$humidity, "x61", "61")
bike[location, ]

datetime        season holiday workingday weather temp  atemp humidity
14177 8/18/2012 21:00 3      0       0          1       27.06 31.06 61      
      windspeed casual registered count sources     
14177 0         90     248        338   www.bing.com

## 3 Converting inputs to data types suitable for analysis

We will now try and convert the columns with wrong data format to a proper data format. There is not always one right data format to put your data in. It depends on what you want to do with it later. Some times it is prefered to have a column values as character strings and other times it is prefered to have them as factors. However, in most cases it make sense to turn dates into a proper data format instead of character strings.

We will start by turning the `humidity` column into a numeric columns as we have now fixed the issue with the non numeric value:

In [11]:
bike$humidity <- as.numeric(bike$humidity)
# bike <- mutate(bike, humidity = as.numeric(humidity))
str(bike)

'data.frame':	17379 obs. of  13 variables:
 $ datetime  : chr  "1/1/2011 0:00" "1/1/2011 1:00" "1/1/2011 2:00" "1/1/2011 3:00" ...
 $ season    : int  1 1 1 1 1 1 1 1 1 1 ...
 $ holiday   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ workingday: int  0 0 0 0 0 0 0 0 0 0 ...
 $ weather   : int  1 1 1 1 1 2 1 1 1 1 ...
 $ temp      : num  9.84 9.02 9.02 9.84 9.84 ...
 $ atemp     : num  14.4 13.6 13.6 14.4 14.4 ...
 $ humidity  : num  81 80 80 75 75 75 80 86 75 76 ...
 $ windspeed : num  0 0 0 0 0 ...
 $ casual    : int  3 8 5 3 0 0 2 1 1 8 ...
 $ registered: int  13 32 27 10 1 1 0 2 7 6 ...
 $ count     : int  16 40 32 13 1 1 2 3 8 14 ...
 $ sources   : chr  "ad campaign" "www.yahoo.com" "www.google.fi" "AD campaign" ...


Next we will look at the `holiday` and `workingday` variables which are numeric, but it is more natural to have them as factors. Here is how to fix that:

In [12]:
bike$holiday <- factor(bike$holiday, levels = c(0, 1), labels = c("no", "yes"))
bike$workingday <- factor(bike$workingday, levels = c(0, 1), labels = c("no", "yes"))
str(bike)

'data.frame':	17379 obs. of  13 variables:
 $ datetime  : chr  "1/1/2011 0:00" "1/1/2011 1:00" "1/1/2011 2:00" "1/1/2011 3:00" ...
 $ season    : int  1 1 1 1 1 1 1 1 1 1 ...
 $ holiday   : Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ workingday: Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ weather   : int  1 1 1 1 1 2 1 1 1 1 ...
 $ temp      : num  9.84 9.02 9.02 9.84 9.84 ...
 $ atemp     : num  14.4 13.6 13.6 14.4 14.4 ...
 $ humidity  : num  81 80 80 75 75 75 80 86 75 76 ...
 $ windspeed : num  0 0 0 0 0 ...
 $ casual    : int  3 8 5 3 0 0 2 1 1 8 ...
 $ registered: int  13 32 27 10 1 1 0 2 7 6 ...
 $ count     : int  16 40 32 13 1 1 2 3 8 14 ...
 $ sources   : chr  "ad campaign" "www.yahoo.com" "www.google.fi" "AD campaign" ...


Note that this code turned `0` into "no" and `1` into "yes". (Underneath R represent "no" as a 1 and "yes" as a 2, which makes it look like we flipped yes and no, but we did not.) To make this decision we need to know that this is correct of course. That is, we need to know something about the data set - that `0` represent "no" and tha `1` represent "yes". Often this kind of information can be found in the *data dictionary*, if the dataset has a such.

In similar manners we can turn the `season` and `weather` columns into factors, which seems to be the right thing to do in this case, as well:

In [13]:
bike$season <- factor(bike$season, levels = c(1, 2, 3, 4), labels = c("spring", "summer", "fall", "winter"), ordered = TRUE)
bike$weather <- factor(bike$weather, levels = c(1, 2, 3, 4), labels = c("clr_part_cloud", "mist_cloudy", "lt_rain_snow", "hvy_rain_snow"), ordered = TRUE)
str(bike)

'data.frame':	17379 obs. of  13 variables:
 $ datetime  : chr  "1/1/2011 0:00" "1/1/2011 1:00" "1/1/2011 2:00" "1/1/2011 3:00" ...
 $ season    : Ord.factor w/ 4 levels "spring"<"summer"<..: 1 1 1 1 1 1 1 1 1 1 ...
 $ holiday   : Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ workingday: Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ weather   : Ord.factor w/ 4 levels "clr_part_cloud"<..: 1 1 1 1 1 2 1 1 1 1 ...
 $ temp      : num  9.84 9.02 9.02 9.84 9.84 ...
 $ atemp     : num  14.4 13.6 13.6 14.4 14.4 ...
 $ humidity  : num  81 80 80 75 75 75 80 86 75 76 ...
 $ windspeed : num  0 0 0 0 0 ...
 $ casual    : int  3 8 5 3 0 0 2 1 1 8 ...
 $ registered: int  13 32 27 10 1 1 0 2 7 6 ...
 $ count     : int  16 40 32 13 1 1 2 3 8 14 ...
 $ sources   : chr  "ad campaign" "www.yahoo.com" "www.google.fi" "AD campaign" ...


**NOTE:** It is not always a good idea to turn character strings into factors. In fact, I will advice not to do it at this stage. You can always do it later, if you realize it will make something easier for you or it is required by other functions you want to use such as a particular function for training a machine learning model, for instance.

We will now finally fix the date format issue with the `datetime` column. To do this we first need to understand what format the `datetime` column is in. Looking at the `str` output above seems to indicate that it is on the format "mm/dd/yyyy hh:mm". The "lubridate" package is a very nice package to work with dates and times. It even has a function to deal with this particular format, namely the function `mdy_hm` function - the name hopefully give away why it is useful in our case! So let us use it to transform the `datetime` column into a proper format:

In [14]:
library(lubridate)
bike$datetime <- mdy_hm(bike$datetime)
str(bike)


Attaching package: 'lubridate'

The following object is masked from 'package:base':

    date



'data.frame':	17379 obs. of  13 variables:
 $ datetime  : POSIXct, format: "2011-01-01 00:00:00" "2011-01-01 01:00:00" ...
 $ season    : Ord.factor w/ 4 levels "spring"<"summer"<..: 1 1 1 1 1 1 1 1 1 1 ...
 $ holiday   : Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ workingday: Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ weather   : Ord.factor w/ 4 levels "clr_part_cloud"<..: 1 1 1 1 1 2 1 1 1 1 ...
 $ temp      : num  9.84 9.02 9.02 9.84 9.84 ...
 $ atemp     : num  14.4 13.6 13.6 14.4 14.4 ...
 $ humidity  : num  81 80 80 75 75 75 80 86 75 76 ...
 $ windspeed : num  0 0 0 0 0 ...
 $ casual    : int  3 8 5 3 0 0 2 1 1 8 ...
 $ registered: int  13 32 27 10 1 1 0 2 7 6 ...
 $ count     : int  16 40 32 13 1 1 2 3 8 14 ...
 $ sources   : chr  "ad campaign" "www.yahoo.com" "www.google.fi" "AD campaign" ...


Note that the `mdy_hm` function turned the datetime column into is something called "POSIXct", which i a common time format in R. (You can google it if you want. We will not go into this i any more details for now.)

## 4 Adapting string variables to a standard

Finally, we will look at adapting sting variables to a standard. The issue here is that sometimes string values are hard to work with and may contain more information than needed. In such cases, with some manipulation we can turn the variable into a factor variable with fewer values, however, values that are still meaningful to us. This is the case for the `source` column, which we will now take a closer look at. First we look at what are all the unique values this column takes:

In [15]:
unique(bike$source)

[1] "ad campaign"      "www.yahoo.com"    "www.google.fi"    "AD campaign"     
 [5] "Twitter"          "www.bing.com"     "www.google.co.uk" "facebook page"   
 [9] "Ad Campaign"      "Twitter    "      NA                 "www.google.com"  
[13] "direct"           "blog"

There are some obvious cleaning we can do here! First of all there are two values for Twitter which should be indentified and there are three values for ad campaing that should probably also be identified. Moreover, we might want to replace `NA` by unknown in this case. The "stringr" package can again help us and we can solve these issues in the following way:

In [16]:
bike$sources <- tolower(bike$sources)
bike$sources <- str_trim(bike$sources)
na_loc <- is.na(bike$sources)
bike$sources[na_loc] <- "unknown"
unique(bike$source)

[1] "ad campaign"      "www.yahoo.com"    "www.google.fi"    "twitter"         
 [5] "www.bing.com"     "www.google.co.uk" "facebook page"    "unknown"         
 [9] "www.google.com"   "direct"           "blog"

This is much better, but we might also want to group all webpages into on category, that is all sources that starts with "www." We can do this using the DataCombine package in the following way:

In [20]:
library(DataCombine)
web_sites <- "(www.[a-z]*.[a-z]*)"
current <- unique(str_subset(bike$sources, web_sites))
replace <- rep("web", length(current))
replacements <- data.frame(from = current, to = replace)
bike <- FindReplace(data = bike, Var = "sources", replacements, from = "from", to = "to", exact = FALSE)
bike$sources <- as.factor(bike$sources)
unique(bike$sources)

Warning message:
"package 'DataCombine' was built under R version 3.6.3"

[1] ad campaign   web           twitter       facebook page unknown      
[6] direct        blog         
Levels: ad campaign blog direct facebook page twitter unknown web

We now have a nice and clean dataset in the right format, which we will use for further analysis in the later lectures. Have a look at it:

In [21]:
str(bike)
bike

'data.frame':	17379 obs. of  13 variables:
 $ datetime  : POSIXct, format: "2011-01-01 00:00:00" "2011-01-01 01:00:00" ...
 $ season    : Ord.factor w/ 4 levels "spring"<"summer"<..: 1 1 1 1 1 1 1 1 1 1 ...
 $ holiday   : Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ workingday: Factor w/ 2 levels "no","yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ weather   : Ord.factor w/ 4 levels "clr_part_cloud"<..: 1 1 1 1 1 2 1 1 1 1 ...
 $ temp      : num  9.84 9.02 9.02 9.84 9.84 ...
 $ atemp     : num  14.4 13.6 13.6 14.4 14.4 ...
 $ humidity  : num  81 80 80 75 75 75 80 86 75 76 ...
 $ windspeed : num  0 0 0 0 0 ...
 $ casual    : int  3 8 5 3 0 0 2 1 1 8 ...
 $ registered: int  13 32 27 10 1 1 0 2 7 6 ...
 $ count     : int  16 40 32 13 1 1 2 3 8 14 ...
 $ sources   : Factor w/ 7 levels "ad campaign",..: 1 7 7 1 5 7 1 7 7 7 ...


datetime            season holiday workingday weather        temp  atemp 
1     2011-01-01 00:00:00 spring no      no         clr_part_cloud  9.84 14.395
2     2011-01-01 01:00:00 spring no      no         clr_part_cloud  9.02 13.635
3     2011-01-01 02:00:00 spring no      no         clr_part_cloud  9.02 13.635
4     2011-01-01 03:00:00 spring no      no         clr_part_cloud  9.84 14.395
5     2011-01-01 04:00:00 spring no      no         clr_part_cloud  9.84 14.395
6     2011-01-01 05:00:00 spring no      no         mist_cloudy     9.84 12.880
7     2011-01-01 06:00:00 spring no      no         clr_part_cloud  9.02 13.635
8     2011-01-01 07:00:00 spring no      no         clr_part_cloud  8.20 12.880
9     2011-01-01 08:00:00 spring no      no         clr_part_cloud  9.84 14.395
10    2011-01-01 09:00:00 spring no      no         clr_part_cloud 13.12 17.425
11    2011-01-01 10:00:00 spring no      no         clr_part_cloud 15.58 19.695
12    2011-01-01 11:00:00 spring no      no         clr_part_cloud 14.76 16.665
13    2011-01-01 12:00:00 spring no      no         clr_part_cloud 17.22 21.210
14    2011-01-01 13:00:00 spring no      no         mist_cloudy    18.86 22.725
15    2011-01-01 14:00:00 spring no      no         mist_cloudy    18.86 22.725
16    2011-01-01 15:00:00 spring no      no         mist_cloudy    18.04 21.970
17    2011-01-01 16:00:00 spring no      no         mist_cloudy    17.22 21.210
18    2011-01-01 17:00:00 spring no      no         mist_cloudy    18.04 21.970
19    2011-01-01 18:00:00 spring no      no         lt_rain_snow   17.22 21.210
20    2011-01-01 19:00:00 spring no      no         lt_rain_snow   17.22 21.210
21    2011-01-01 20:00:00 spring no      no         mist_cloudy    16.40 20.455
22    2011-01-01 21:00:00 spring no      no         mist_cloudy    16.40 20.455
23    2011-01-01 22:00:00 spring no      no         mist_cloudy    16.40 20.455
24    2011-01-01 23:00:00 spring no      no         mist_cloudy    18.86 22.725
25    2011-01-02 00:00:00 spring no      no         mist_cloudy    18.86 22.725
26    2011-01-02 01:00:00 spring no      no         mist_cloudy    18.04 21.970
27    2011-01-02 02:00:00 spring no      no         mist_cloudy    17.22 21.210
28    2011-01-02 03:00:00 spring no      no         mist_cloudy    18.86 22.725
29    2011-01-02 04:00:00 spring no      no         mist_cloudy    18.86 22.725
30    2011-01-02 06:00:00 spring no      no         lt_rain_snow   17.22 21.210
...   ...                 ...    ...     ...        ...            ...   ...   
17350 2012-12-30 18:00:00 spring no      no         mist_cloudy     9.84 10.605
17351 2012-12-30 19:00:00 spring no      no         clr_part_cloud 13.94 18.180
17352 2012-12-30 20:00:00 spring no      no         clr_part_cloud  9.02  9.850
17353 2012-12-30 21:00:00 spring no      no         clr_part_cloud  8.20 10.605
17354 2012-12-30 22:00:00 spring no      no         clr_part_cloud  8.20  9.850
17355 2012-12-30 23:00:00 spring no      no         clr_part_cloud  8.20  9.850
17356 2012-12-31 00:00:00 spring no      yes        clr_part_cloud  7.38  9.090
17357 2012-12-31 01:00:00 spring no      yes        clr_part_cloud  7.38  9.090
17358 2012-12-31 02:00:00 spring no      yes        clr_part_cloud  6.56  8.335
17359 2012-12-31 03:00:00 spring no      yes        clr_part_cloud  6.56  9.090
17360 2012-12-31 04:00:00 spring no      yes        clr_part_cloud  5.74  8.335
17361 2012-12-31 05:00:00 spring no      yes        clr_part_cloud  6.56  7.575
17362 2012-12-31 06:00:00 spring no      yes        clr_part_cloud  6.56  8.335
17363 2012-12-31 07:00:00 spring no      yes        clr_part_cloud  6.56  9.090
17364 2012-12-31 08:00:00 spring no      yes        clr_part_cloud  5.74  7.575
17365 2012-12-31 09:00:00 spring no      yes        mist_cloudy     7.38 10.605
17366 2012-12-31 10:00:00 spring no      yes        mist_cloudy     8.20 10.605
17367 2012-12-31 11:00:00 spring no      yes        mist_cloudy     9.02 11.365
17368 

# Exercise

In this exercise, we will load in another dataset from Google Analytics, clean it and reshape it. 

First, download the sheet "Dataset2" from the dataset "Webanalytics_data_example2.xlsx" and save in a variable called `gadata`. This is website data about users for the first 6 months of 2019 divided by age group.

*Hint: Think about what package and data loading function that will be useful here.*

In [26]:
library(readxl)
gadata <- read_excel("Webanalytics_data_example2.xlsx", skip = 1, sheet = "Dataset2")
head(gadata)

New names:
* Users -> Users...2
* Users -> Users...3
* Users -> Users...4
* Users -> Users...5
* Users -> Users...6
* ...


Age   Users...2 Users...3 Users...4 Users...5 Users...6 Users...7 Users...8
1 25-34 80304     720       722       713       732       682       644      
2 35-44 37631     340       321       281       310       339       300      
3 18-24 31384     227       250       218       228       208       230      
4 45-54 12819     132       121       112        85        93       103      
5 55-64  4393      40        30        41        29        27        25      
6 65+    2009      18        17        24         0        20        17      
  Users...9 Users...10 ... Users...174 Users...175 Users...176 Users...177
1 636       582        ... 229         230         233         207        
2 286       265        ...  95         120          93         112        
3 187       236        ... 125         118         126         122        
4  96        86        ...  45          45          53          50        
5  23        41        ...  14          12          20          22        
6  16        16        ...  18           0           0          11        
  Users...178 Users...179 Users...180 Users...181 Users...182 Users...183
1 213         224         172         234         204         187        
2 127         102         129         114         113         101        
3 108         105         120          99          84          78        
4  43          43          55          40          39          33        
5  19          15          23          11          13          15        
6  14          15           0           0           0          11

If you managed to load in the data it probably looks like this:
![gadata](gadata1.png)
It looks a bit strange doesn't it? It certainly does not look like tidy data. We want to turn it into tidy data that looks like this:
![gadata2](gadata2.png)

Doing this cleaning will require several steps, so we will go through them one by one.

The first thing we will do is to remove the second column titled "Total" as it is just a column with totals, which is not really an observation or an attribute according to the concept of tidy data. So remove the second "Total" column.

*Hint: You can use the `select` function. Try to figure out how `-` can be used in connection with the `select` function.*

In [34]:
names(gadata)
#hmmm can't find this "total" column

[1] "Age"         "Users...2"   "Users...3"   "Users...4"   "Users...5"  
  [6] "Users...6"   "Users...7"   "Users...8"   "Users...9"   "Users...10" 
 [11] "Users...11"  "Users...12"  "Users...13"  "Users...14"  "Users...15" 
 [16] "Users...16"  "Users...17"  "Users...18"  "Users...19"  "Users...20" 
 [21] "Users...21"  "Users...22"  "Users...23"  "Users...24"  "Users...25" 
 [26] "Users...26"  "Users...27"  "Users...28"  "Users...29"  "Users...30" 
 [31] "Users...31"  "Users...32"  "Users...33"  "Users...34"  "Users...35" 
 [36] "Users...36"  "Users...37"  "Users...38"  "Users...39"  "Users...40" 
 [41] "Users...41"  "Users...42"  "Users...43"  "Users...44"  "Users...45" 
 [46] "Users...46"  "Users...47"  "Users...48"  "Users...49"  "Users...50" 
 [51] "Users...51"  "Users...52"  "Users...53"  "Users...54"  "Users...55" 
 [56] "Users...56"  "Users...57"  "Users...58"  "Users...59"  "Users...60" 
 [61] "Users...61"  "Users...62"  "Users...63"  "Users...64"  "Users...65" 
 [66] "Users...66"  "Users...67"  "Users...68"  "Users...69"  "Users...70" 
 [71] "Users...71"  "Users...72"  "Users...73"  "Users...74"  "Users...75" 
 [76] "Users...76"  "Users...77"  "Users...78"  "Users...79"  "Users...80" 
 [81] "Users...81"  "Users...82"  "Users...83"  "Users...84"  "Users...85" 
 [86] "Users...86"  "Users...87"  "Users...88"  "Users...89"  "Users...90" 
 [91] "Users...91"  "Users...92"  "Users...93"  "Users...94"  "Users...95" 
 [96] "Users...96"  "Users...97"  "Users...98"  "Users...99"  "Users...100"
[101] "Users...101" "Users...102" "Users...103" "Users...104" "Users...105"
[106] "Users...106" "Users...107" "Users...108" "Users...109" "Users...110"
[111] "Users...111" "Users...112" "Users...113" "Users...114" "Users...115"
[116] "Users...116" "Users...117" "Users...118" "Users...119" "Users...120"
[121] "Users...121" "Users...122" "Users...123" "Users...124" "Users...125"
[126] "Users...126" "Users...127" "Users...128" "Users...129" "Users...130"
[131] "Users...131" "Users...132" "Users...133" "Users...134" "Users...135"
[136] "Users...136" "Users...137" "Users...138" "Users...139" "Users...140"
[141] "Users...141" "Users...142" "Users...143" "Users...144" "Users...145"
[146] "Users...146" "Users...147" "Users...148" "Users...149" "Users...150"
[151] "Users...151" "Users...152" "Users...153" "Users...154" "Users...155"
[156] "Users...156" "Users...157" "Users...158" "Users...159" "Users...160"
[161] "Users...161" "Users...162" "Users...163" "Users...164" "Users...165"
[166] "Users...166" "Users...167" "Users...168" "Users...169" "Users...170"
[171] "Users...171" "Users...172" "Users...173" "Users...174" "Users...175"
[176] "Users...176" "Users...177" "Users...178" "Users...179" "Users...180"
[181] "Users...181" "Users...182" "Users...183"

The second thing we will do is to rename the first column from "...1" to "Age". So go ahead and do this.

*Hint: Use the `rename` function from the dplyr package.*

The next thing we will do is to remove the first row as it just contains the word "Age" or "Users". (Note that what looks like the first row is not a row, but names of columns.) 

*Hint: This can be done in several ways. Try either to subset the data with the "[ , ]", or use the `filter` function.*

Next, we will remove the last rows as it contains totals as well.

*Hint: You can use the same technique as in the previous question.*

Now your data frame should look like this:
![gadata3](gadata3.png)
We will turn all the columns execept `Age` column into one column with the gather function, so that our data will look like this:
![gadata4](gadata4.png)

Go ahead and do this!

*Hint: Use the `gather` function. Call the key column "Date" and the value column "Users". To select every column except the age column as the last arguments there are different options, but try out the `starts_with`.* 

We will now spread out the data again such that the age groups are columns. 

*Hint: Use the spread function with the column `Age` as key and the column `Users` as value.

We will now turn the numbers in the "Date" column into actual dates. Before this, we will arrange the "Date" columns such that the lowest date is first. (The dates are numbered and represent days since "2019-01-01".)

*Hint: Use the arrange function.*

We will now put in the dates instead of the numbers. To do this we will first generate the relevant dates. This can be done with the following command

In [ ]:
dates <- seq(from = as.Date("2019-01-01"), to = as.Date("2019-06-30"), by = "day")
dates

We can now put then in instead of the numbers in the `Date` column. Do this.

Your data should now look like this:
![gadata5](gadata5.png)

Now we will gather the data again to get in a tidy format. 

*Hint: Use the `gather` function, "Age" as key and "Users" as value. To select the age group columns you can use `"18-24":"65+"`*

Finally, we will make sure that the `Age` column is factor and that the column `Users` is numeric. So go ahead and make this transformation.

*Hint: use the functions `as.factor` and `as.numeric`.*

If you managed to get the data in the right format the code below should give you a nice plot

In [ ]:
ggplot(data = gadata) + geom_line(mapping = aes(x = Date, y = Users, color = Age))